Auto-generated Semantic Conventions

joaopgrassi opened the pull request:
Fixes #1658.

## Changes

As discussed in #1658, this PR introduces a new package `OpenTelemetry.SemanticConventions` which aims to be the home of the auto-generated semantic convention code files. It uses the [Semantic Convention Generator](https://github.com/open-telemetry/build-tools/blob/main/semantic-conventions/README.md), similar as it's done in other languages.

Few things I noticed while doing this, which probably need some discussion and agreement:

- The existing `SemanticConventions.cs` class. Looking at the other languages, they call it:
  - Java: `SemanticAttributes`
  - Python: `SpanAttributes`
  - JS: `SemanticAttributes`

- The existing `ResourceSemanticConventions` class. The other languages call it pretty much `ResourceAttributes`. But this name conflicts with `ResourceAttributes` from `System.Reflection`, so probably not a good name to use.

What should we use here? I went with `TraceSemanticConventions`, which is not similar to any of the above, but is similar to what we have with the resource one.

For now, I went with this:

- `SemanticConventions` -> `TraceSemanticConventions`. Inside namespace `OpenTelemetry.SemanticConventions.Trace`
- `ResourceSemanticConventions` -> same name. Inside namespace `OpenTelemetry.SemanticConventions.Resource`

The code files are generated via a script, manually for now. Once the naming above is agreed, I can add this new project to all others in the solution and start renaming and see if all works. Since they were internal files, I think it shouldn't be a problem?

pcwiese commented:
@cijothomas @reyang for guidance on naming here.

bruno-garcia commented on src/OpenTelemetry.SemanticConventions/Resource/ResourceSemanticConventions.cs:
```diff
@@ -0,0 +1,714 @@
+        /// <summary>
+        /// Name of the cloud provider.
+        /// </summary>
+        public const string AttributeCloudProvider = "cloud.provider";
```
What about an inner class `Attribute` and having these generated under it?
So we'd access it through:

`ResourceSemanticConventions.Attribute.CloudProvider`

joaopgrassi replied:
I like the idea. Let's see what the others say then I can change.

bruno-garcia commented on src/OpenTelemetry.SemanticConventions/OpenTelemetry.SemanticConventions.csproj:
```diff
@@ -0,0 +1,14 @@
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <NoWarn>$(NoWarn),1591,CS0618</NoWarn>
```
nit: Not sure if this is just a copy from other projects but I usually suggest adding a comment above explaining what each ignored warning is about and why it's ignored.

joaopgrassi replied:
Yeah it was a copy and paste. I removed it since `1591` is for xml docs on public members, which should be there always. `CS0618` was for Obsolete. This one I don't know if we should keep, since the attributes can be marked obsolete in the spec.. but since here it just defines them and don't reference I'm not sure.

bruno-garcia commented on src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/generate.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Do we usually have a `.ps1` version for each `.sh` script or specially these days with `wsl` it doesn't matter?
Definitely easier to maintain only one script.

joaopgrassi replied:
It's there now :). Tested it produces the same result as the other.

bruno-garcia reviewed (commented):
I think this is great stuff! Left some suggestions

bruno-garcia commented:
Also, this seems pretty mature already, any particular reason for it to be a draft?
You'll get the most reviews from folks once you drop the draft flag

joaopgrassi commented:
@bruno-garcia I'll go through the comments, thanks! 

> Also, this seems pretty mature already, any particular reason for it to be a draft? You'll get the most reviews from folks once you drop the draft flag

It's mostly because I have all these naming stuff that I'm really not sure, so there's still quite some stuff to do. Once the namings are defined I'll have to go and change everywhere and see if everything still "works". 

Then the idea is to merge this, and think how we will publish this package. Only then we'll use it in all the projects that use the internal class before. At least that's what @pcwiese  and I discussed in the linked issue and in Slack.

joaopgrassi commented:
Publishing it now, so maybe we can get some feedback. As @bruno-garcia said I think now it's in a good state, just need to define the small details.

cijothomas commented on src/OpenTelemetry.SemanticConventions/OpenTelemetry.SemanticConventions.csproj:
```diff
@@ -0,0 +1,10 @@
+    <TargetFrameworks>netstandard2.0;net452</TargetFrameworks>
+    <Description>OpenTelemetry Semantic Conventions</Description>
+    <PackageTags>$(PackageTags);distributed-tracing</PackageTags>
+    <MinVerTagPrefix>core-</MinVerTagPrefix>
```
lets remove "core-" now, else this will be versioned same as core packages, but as per spec, semantic conventions are not still marked experimental, so it should be released as pre-release.

joaopgrassi replied:
Ah I see. So, I guess just remove the MinVer tag altogether here and when we release this we use the pre-release tag?

cijothomas commented:
> For now, I went with this:
> 
> SemanticConventions -> TraceSemanticConventions. Inside namespace OpenTelemetry.SemanticConventions.Trace
> ResourceSemanticConventions -> same name. Inside namespace OpenTelemetry.SemanticConventions.Resource

TraceSemanticConventions inside Otel.SemanticConventions.Trace seems bit redundant. We may name the namespace as `OpenTelemetry.SemanticConventions`, and classes `TraceSemanticConventions`, `MetricsSemanticConventions` and `ResourceSemanticConventions`

cijothomas commented:
> > For now, I went with this:
> > SemanticConventions -> TraceSemanticConventions. Inside namespace OpenTelemetry.SemanticConventions.Trace
> > ResourceSemanticConventions -> same name. Inside namespace OpenTelemetry.SemanticConventions.Resource
> 
> TraceSemanticConventions inside Otel.SemanticConventions.Trace seems bit redundant. We may name the namespace as `OpenTelemetry.SemanticConventions`, and classes `TraceSemanticConventions`, `MetricsSemanticConventions` and `ResourceSemanticConventions`

Alternately, if we re-use existing namespaces `OpenTelemetry.Trace`, `Opentelemetry.Resource` etc we can name classes as `TraceSemanticConventions`, `MetricsSemanticConventions` and `ResourceSemanticConventions`. This will avoid requires users to import more namespaces.

joaopgrassi commented:
@cijothomas thanks for the review! I changed the namespace as your last comment. 

What do you think regarding @bruno-garcia suggestion on having an inner class for the attributes? 

> What about an inner class Attribute and having these generated under it?
So we'd access it through:

> ResourceSemanticConventions.Attribute.CloudProvider

And how do you think we can release this? Via CI? manual? @pcwiese had some comments in the original issue ticket... also how we will use this package later on the other projects.

joaopgrassi commented:
Something else I forgot: Should I add the files for  `.publicApi` for this as well?

joaopgrassi commented:
I added the `.publicApi` files. The `sanitycheck/encoding` CI task is failing due to a non-ascii char in one of the xml docs. It comes from here: https://github.com/open-telemetry/opentelemetry-specification/blob/d271b73d6b274bfd51eae9ed813586f0322342b7/semantic_conventions/trace/database.yaml#L65.

Not sure what to do about it.. any suggestions?

moonheart commented:
> I added the `.publicApi` files. The `sanitycheck/encoding` CI task is failing due to a non-ascii char in one of the xml docs. It comes from here: [open-telemetry/opentelemetry-specification@`d271b73`/semantic_conventions/trace/database.yaml#L65](https://github.com/open-telemetry/opentelemetry-specification/blob/d271b73d6b274bfd51eae9ed813586f0322342b7/semantic_conventions/trace/database.yaml#L65).
> 
> Not sure what to do about it.. any suggestions?

'InterSystems Caché' `é` is the non-ascii char.

cijothomas commented on src/OpenTelemetry.SemanticConventions/OpenTelemetry.SemanticConventions.csproj:
```diff
@@ -0,0 +1,9 @@
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;net452</TargetFrameworks>
+    <Description>OpenTelemetry Semantic Conventions</Description>
+    <PackageTags>$(PackageTags);semantic-convetions</PackageTags>
```
```suggestion
    <PackageTags>$(PackageTags);semantic-conventions</PackageTags>
```

joaopgrassi commented:
> 'InterSystems Caché' `é` is the non-ascii char.
> 
> You can check it [here](https://pages.cs.wisc.edu/~markm/ascii.html).

Yes, that's what I mentioned above. It is coming from the yaml file. None of the alternatives I thought seem to be good. 

- Send a PR to remove the ascii on the source probably not viable
- Removing the xml comments for the auto-generated code files also not great

Thought of maybe defining some mapping and remove it during the code generation. For this case, substitute `é` with `e`, but also sounds like a hack. If tomorrow new ones are introduce, we need to change this again. 

To me, the best option would be to disable the CI check for this package, but that is not up to me so waiting on others for directions.

joaopgrassi commented on src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/.gitignore:
```diff
@@ -0,0 +1 @@
+opentelemetry-specification/
```
I guess I could do that as well. I feel that it's nice that all is close together (scripts, templates and the spec) and the script is straightforward. If I had to checkout somewhere else than things will get a bit more convoluted I guess.

Oberon00 commented on src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/templates/SemanticConventions.cs.j2:
```diff
@@ -0,0 +1,101 @@
+{%- macro print_value(type, value) -%}
+  {{ "\"" if type == "string"}}{{value}}{{ "\"" if type == "string"}}
```
Oops. But you could use `json.dumps(name)`

joaopgrassi replied:
I think I'll keep like this. I like that it's clear what it does even if you are not familiar with Python.

Oberon00 commented on src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/templates/SemanticConventions.cs.j2:
```diff
@@ -0,0 +1,101 @@
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+        public const string AttributeDbInstance = "db.instance";
+
+        public const string AttributeExceptionEventName = "exception";
```
Probably `.GROUP_TYPE_NAME` can be accessed on the semantic convention instance.

joaopgrassi replied:
Managed to do it. I have now included `prefixes` and `event`. E.g.
```csharp
public static readonly string PrefixHttp = "http";
...
public static readonly string EventException = "exception";
```

I followed the `Attribute` prefix, and added `Prefix` and `Event` so at least it's consistent. What do you think?

joaopgrassi commented on src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/templates/SemanticConventions.cs.j2:
```diff
@@ -0,0 +1,101 @@
+
+{%- if class == "TraceSemanticConventions" %}
+        // Manually defined and not YET in the YAML
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
```
Done. Not sure what is that used for so I simply added a very generic message.

Oberon00 commented on src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/templates/SemanticConventions.cs.j2:
```diff
@@ -0,0 +1,117 @@
+{%- endmacro %}
+
+{%- macro format_remarks(text) -%}
+{%- set notes = '\n ///'.join(text.splitlines()).encode('ascii', 'xmlcharrefreplace').decode() -%}
+{{notes}}
```
```suggestion
{{ '\n ///'.join(text.splitlines()).encode('ascii', 'xmlcharrefreplace').decode() }}
```

I think this should also work.

cijothomas commented:
@joaopgrassi do you have time to resurrect this?

joaopgrassi commented:
Hey @nahk-ivanov @cijothomas I have been working on and off, to try to keep it up with the conventions. I have some problems right now with the semconv tooling  (jinja generation) + dotnet format. I will try to work on it in the upcoming days/weeks. I very much want this to get merged 😊

nahk-ivanov commented:
Thanks @joaopgrassi, the World will remember! 😉 

I used your generated files and it works for the most part with two things I've changed:
1) Remove `GROUP_TYPE_NAME` outside of XMLDOC tags: [here](https://github.com/joaopgrassi/opentelemetry-dotnet/blob/5ee7692e3a19813742965f85e5a35173a44c7e0f/src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/templates/SemanticConventions.cs.j2#L79) and [here](https://github.com/joaopgrassi/opentelemetry-dotnet/blob/5ee7692e3a19813742965f85e5a35173a44c7e0f/src/OpenTelemetry.SemanticConventions/scripts/semantic-convetion/templates/SemanticConventions.cs.j2#L87) in the template
2) I'm not sure why prefixes and event names are `static`, while attribute names are `const` -- yeah, public compile-time constants should be avoided, but I think people mostly don't care at this point and in this case it would be beneficial to have everything as `const`, as people will be able to use them declaratively (in attributes, etc.).

In regards to OpenTelemetry itself - I don't really understand why common attributes are redefined for different types of telemetry. For example, both traces and metrics redefine common HTTP attributes (`http.method`, `http.scheme`, etc.) for respective semantic conventions. I think these should be orthogonal to the types of telemetry, but today they are defined explicitly [within traces](https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/http.yaml), for example. Once metrics are added, we will have same attributes in two places for no apparent reason.

Let me know if/how I can help you with this!

joaopgrassi commented:
Hi @nahk-ivanov 

> Remove GROUP_TYPE_NAME outside of XMLDOC

Good catch, some left over when I was trying stuff out. Removed

> I'm not sure why prefixes and event names are static, while attribute names are const

Yeah this is also some overlook on my part as well. I'd be more inclined in changing all to `static readonly` for the reasons you pointed. I'm not sure if there's even a use case for using the semconv attributes in c# attributes. 

> I'm also not sure why generated files are in source control, we should gitignore them, me thinks. Build should generate them before packaging.

I think the opposite. The process we are aiming as far as I remember is manual at first. Someone would bump the semconv version and run the script, check things and then commit. Code generation like this is always a bit tricky so I wouldn't do anything automatic like that. Maybe when things are more stable we can move to such approach. Having it in source control also makes it easy to know what changed between versions. Since this PR was open, I had to keep updating it with the new versions and several it broke because things like new lines, special characters were introduced that the template wasn't prepared for.  


> In regards to OpenTelemetry itself - I don't really understand why common attributes are redefined for different types of

This is a known issue. There was an intention to refactor it, pretty much to achieve what you described but it got stale. It's not a trivial change and involves many things I guess. (https://github.com/open-telemetry/opentelemetry-specification/pull/1977)

As for the status of the PR. I opened a PR on the build-tools, in order to be able to turn off the whitespace behavior in Jinja. Now spaces in the template actually means a space/newline is rendered and it's MUCH easier to reason about. Before I had to use this `{%- %+` syntax in the template and I couldn't fix several cases without huge hacks. Like I would end up with newlines, empty spaces.. just a complete mess. I thought of leaving it and relying on `dotnet format` to fix it but that also added more issues.. with the latest update being it couldn't "merge" things and it just left the code broken. 

Now with the changed template relying in https://github.com/open-telemetry/build-tools/pull/101, the `.cs` files are pretty much impeccable, apart from a small case with multi-line `<remarks>` but I think that's really minor and negligible.

nahk-ivanov commented:
This is awesome! Merge and publish to nuget.org like right now! 😍 

In regards to
> I think the opposite. The process we are aiming as far as I remember is manual at first. Someone would bump the semconv version and run the script, check things and then commit. Code generation like this is always a bit tricky so I wouldn't do anything automatic like that.

I don't know... In most of the cases we should be able to automate it anyway, so steps like clone the repo and commit just add to the process that should be one-click. If there is something that is utterly broken - build will fail and we will clone the repo at that point to fix it. If there is some new stuff added to the yaml that we don't support yet in here - then we clone the repo, update the template and rerun the build. It should get new build/revision/patch/(whatever they decide to call it today), which will have that additional stuff.

Regardless, waiting for this for over a year I'd be fine with anything that doesn't require me to copy-paste. :)

cartermp commented:
Looks like the only thing red is a sanity check that is detecting some extraneous spaces in XML doc comments. Is that required to be ale to push this through? I'd be happy to followup with a PR that just manually removes them for now, then figure out the source of it (i.e., is it in the generator tool, or some older version of the yaml specs this is generated from, etc.)

joaopgrassi commented:
Hi @cartermp 
In general the PR is almost ready. All the spaces, and formatting is really tricky to get it right. I was doing a best effort in the Jinja template and then let `dotnet format` fix the rest.. but I ran into several problems with dotnet format messing up the file so I aborted that approach. 

The thing I mainly struggled was how the whitespace rendering works in Jinja. I since then submitted a PR to allow customizing it in the semconv generator so now the template on our side is "easier to reason about". 

I still need to find out where the extra spaces here are coming I assume they are mostly coming from the Faas resource semconv. There's also `< >` characters in the conventions `brief` that breakes the xml formatting in .NET. https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/resource/faas.yaml#L24

Could send a PR to change that to use SUBSCIPTION_GUID instead of <SUBSCIPTION_GUID> and use ` to show it as `<c></c>`. The problem with this approach is that as soon as someone else adds a new conventions using the same thing, our generation is broken, so I don't think that's a good idea. I feel we need to do some magic in the jinja template and escape such chars. 

I don't have many cycles to work on it right now, but I will see what I can investigate and fix in the upcoming weeks.

cartermp commented:
Makes sense! I'll defer to maintainers here on how to best proceed. It's not strictly urgent (just a docs update), but if we get an indication that the time to update w.r.t. the latest spec will be longer for reasons that can't be controlled here, I'd be happy to help with anything needed to get it merged+released as-is (and then documented).

cijothomas commented on src/OpenTelemetry.SemanticConventions/OpenTelemetry.SemanticConventions.csproj:
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
```
we can do net462 and higher only.
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/OpenTelemetry.Api.csproj#L4

joaopgrassi replied:
@cijothomas  sorry I didn't have much time to look into these. Busy recent days.. could you please re-open this? I will try to get around this on the upcoming days/weekend.

joaopgrassi replied:
Done. Targeting now `net462`, updated publicApis.

joaopgrassi commented:
@cijothomas I think this is ready for a final review. I updated it with the latest spec release (1.13), ironed out a few warnings and re-generated the public api files. The build seems to fail in Windows  for net6.0, not sure why. I can't find a way to re-try failed jobs.

cijothomas reviewed (approved):
thanks for getting this ready!
Lets try and incorporate this into existing projects one by one, and flush out any issues in that process, before officially releasing this as a nuget.
